### DISCUSSION TITLE
iTwinJs Error: Lock Not Held: error updating element: exclusive lock not held on element for update

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#31] [Q&A] [answered]
Hello I work on developing the FutureOn - iTwin connector.
Running a connector synchronisation is failing with the error:
```
backend     'Lock Not Held: error updating element: exclusive lock not held on element for update (id=0xe0000000005)\n' +
backend     '    at ServerBasedLocks.checkExclusiveLock (/app/node_modules/@itwin/core-backend/lib/cjs/ServerBasedLocks.js:160:19)\n' +
backend     '    at PhysicalPartition.onUpdate (/app/node_modules/@itwin/core-backend/lib/cjs/Element.js:82:22)\n' +
backend     '    at Elements.updateElement (/app/node_modules/@itwin/core-backend/lib/cjs/IModelDb.js:1565:39)\n' +
backend     '    at FutureOnBridge.updateElement (/app/node_modules/@xvisionas/fieldtwin-itwin-connector/lib/src/FutureOnBridge.js:1267:43)\n' +
backend     '    at FutureOnBridge.getPhysicalPartition (/app/node_modules/@xvisionas/fieldtwin-itwin-connector/lib/src/FutureOnBridge.js:349:18)\n' +
backend     '    at FutureOnBridge.removePhysicalModel (/app/node_modules/@xvisionas/fieldtwin-itwin-connector/lib/src/FutureOnBridge.js:408:33)\n' +
backend     '    at FutureOnBridge.removeBathyModel (/app/node_modules/@xvisionas/fieldtwin-itwin-connector/lib/src/FutureOnBridge.js:1193:14)\n' +
backend     '    at FutureOnBridge.updateExistingData (/app/node_modules/@xvisionas/fieldtwin-itwin-connector/lib/src/FutureOnBridge.js:298:28)\n' +
backend     '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
backend     '    at async /app/node_modules/@itwin/connector-framework/lib/src/ConnectorRunner.js:201:13
```

Can you help to provide more information on what could be causing this error?

**Comment by PeterJBassett** [comment]
The connector version we are using is:
```
    "@itwin/connector-framework": "^2.0.5",
    "@itwin/core-backend": "^4.4.0",
    "@itwin/core-bentley": "^4.4.0",
    "@itwin/core-common": "^4.4.0",
    "@itwin/core-geometry": "^4.4.0",
    "@itwin/ecschema-metadata": "^4.4.0",
    "@itwin/imodels-access-backend": "^4.1.0",
    "@itwin/imodels-client-authoring": "^4.1.8",
```

**Comment by pmconne** [comment]
The error is a result of using [pessimistic concurrency](https://www.itwinjs.org/learning/backend/concurrencycontrol/) and failing to obtain a lock on the element you're trying to modify, before trying to modify it.

Is this a new problem? Did something change (dependency versions, your code, etc) before it began occurring? Was the physical partition you are attempting to update created by a previous run of your connector, or by something else?

Presumably your connector creates some Subject, and then puts all of the models+elements it creates under that subject's hierarchy. You should be able to use [iModelDb.locks.acquireLocks](https://www.itwinjs.org/reference/core-backend/imodels/lockcontrol/acquirelocks/) to exclusively lock your root subject, which implicitly locks everything underneath that subject too.

**Reply by PeterJBassett** [reply]
Hello @pmconne Thank you for your reply. 
No, nothing has changed in the connector code. The connector was run successfully a few hours before the failure.

**Reply by PeterJBassett** [reply]
Hello @pmconne @jffmarker 
I have tried to reproduce the problem by syncing to my own iModel, but have not been able to reproduce the problem.

It seems to be specific to the iModel that Shell are trying to sync. There seems to be some state set in the iModel that is preventing the lock being set.

The iModel that is failing is:
(This is owned by Shell, and I don't have any access)
```
iModelProjectId = 2474f1ed-9866-4ecd-b1f8-a692204119a7
iModelId = 2a2d3cfc-21cb-448a-a702-60b69d343961
```

**Reply by pmconne** [reply]
As described in the concurrency article that I linked, the concurrency mode is a property of the iModel. Shell's iModel presumably has pessimistic locking enabled, while your iModel does not. Did you try my suggested change?

**Reply by PeterJBassett** [reply]
Hello @pmconne thanks for your reply.  I don't understand how this explains the problem? The connector has been working correctly for a long time. It is currently working correctly in Shell Prod, but is failing only for this specific iModel in Shell QA.

I haven't tried the suggested change. This would require a code change to the connector, and then deployment to Shell. I am using the iTwin Connector Framework which (in my understanding) manages the locks internally.
https://www.itwinjs.org/reference/connector-framework/

**Reply by pmconne** [reply]
Check the be_Local tables of one iModel that works, and one that fails. Report back with which one(s) contain a "NoLocking" entry.

**Reply by PeterJBassett** [reply]
Hello @pmconne I can't do that, as I don't have access to Shell's iModels. It would need someone in Bentley to do this; who does have access.

**Reply by pmconne** [reply]
We don't have access to user's iModels either. Ask the user?

**Reply by PeterJBassett** [reply, marked as answer]
Hello @pmconne I have managed to track down the problem in my connector code which appears to be a timing problem acquiring the iModel lock.

So this can be closed.
